nexus/db-queries/src/db/datastore/db_metadata.rs

```diff
@@ -13,13 +13,93 @@ use async_bb8_diesel::{AsyncRunQueryDsl, AsyncSimpleConnection};
 use chrono::Utc;
 use diesel::prelude::*;
 use nexus_db_model::AllSchemaVersions;
+use nexus_db_model::SchemaUpgradeStep;
 use nexus_db_model::SchemaVersion;
 use nexus_db_model::EARLIEST_SUPPORTED_VERSION;
 use omicron_common::api::external::Error;
 use omicron_common::api::external::SemverVersion;
 use slog::{error, info, o, Logger};
+use std::ops::Bound;
 use std::str::FromStr;
 
+// A SchemaVersion which uses a pre-release value to indicate
+// "incremental progress".
+//
+// Although we organize our schema changes into higher-level versions,
+// transaction semantics of CockroachDB do not allow multiple DDL statements
+// to be applied together. As a result, an upgrade from "1.0.0" to "2.0.0"
+// may involve multiple smaller "step" versions, each separated by
+// an auto-generated pre-release version.
+#[derive(Clone)]
+struct StepSemverVersion {
+    version: SemverVersion,
+    i: usize,
+}
+
+impl StepSemverVersion {
+    // Durably store our progress in the target version as the
+    // pre-release extension of "target_version".
+    //
+    // Note that the format of "step.{i}" is load-bearing, see:
+    // https://docs.rs/semver/1.0.22/semver/struct.Prerelease.html#examples
+    //
+    // By using "dot number" notation, we can order the pre-release
+    // steps, which matters while comparing their order.
+    fn new(target_version: &SemverVersion, i: usize) -> anyhow::Result<Self> {
+        let mut target_step_version = target_version.clone();
+        target_step_version.0.pre =
+            semver::Prerelease::new(&format!("step.{i}"))
+                .context("Cannot parse step as semver pre-release")?;
+        Ok(Self { version: target_step_version, i })
+    }
+
+    // Drops the pre-release information about this target version.
+    //
+    // This is the version we are upgrading to, using these incremental steps.
+    fn without_prerelease(&self) -> SemverVersion {
+        let mut target_version = self.version.clone();
+        target_version.0.pre = semver::Prerelease::EMPTY;
+        target_version
+    }
+
+    fn previous(&self) -> anyhow::Result<Self> {
```

Comment:
Tiny nit - I'd maybe make this return `Option<Self>` directly, since the only caller of it immediately calls `.ok()` to do that conversion.

Reply:
Sounds good, updated